fix(isMatch): handle primitive targets with object source patterns correctly

sen2y opened the pull request:
# Fix: Correct isMatch primitive target vs object pattern matching

Fixes #1399

---

## 🐛 Problem

Two incorrect matches versus Lodash were identified.

### Bug 1: Top-level primitive target vs object source (with undefined)

Case:

```ts
isMatch('bar', { anyKey: undefined });
```

Observed behavior before fix:

- Lodash: false ✅
- ES-Toolkit (before): true ❌

Root cause:

- Accessing a non-existent key on a primitive target yields `undefined`:

```ts
'bar'['anyKey']; // -> undefined
source['anyKey']; // -> undefined
```

These two `undefined` values accidentally compare as equal, yielding a false positive.

### Bug 2: Nested level primitive vs empty object

Case:

```ts
isMatch({ value: 'bar' }, { value: {} });
```

Observed behavior before fix:

- Lodash: false ✅
- ES-Toolkit (before): true ❌

Why it is confusing:

- Top-level: `isMatch('bar', {})` should be true (empty object = no constraints)
- Nested: `isMatch('bar', {})` must be false (type mismatch vs object pattern)

Lodash intent:

- Top-level empty object is permissive (matches anything)
- Nested empty object still enforces object type (primitive must not match object)

ES-Toolkit previously treated empty objects as truthy matches at all levels.

---

## 🔧 Solution

1. Reject object-pattern matching against primitive targets

```ts
// If source expects object keys, target must be an object
if (!isObject(target)) {
  return false;
}
```

Fixes Bug 1:

```ts
isMatch('bar', { anyKey: undefined }); // false ✅
isMatch(123, { x: 1 }); // false ✅
```

2. Distinguish top-level vs nested level for empty-object source using stack size

```ts
if (keys.length === 0) {
  // Only at nested levels, primitives must not match {}
  if (stack && stack.size > 0 && !isObject(target)) {
    return false;
  }
  return true;
}
```

Rationale:

- At the top level (`stack.size === 0`), `{}` means "no constraints" -> allow any target
- At nested levels (`stack.size > 0`), `{}` still implies object type -> primitives must fail

Implementation detail:

- `isMatchWith` creates a new stack (`new Map()`) at the entry point.
- On recursion, objects are pushed to the stack (`stack.set(...)`), so `stack.size` naturally reflects depth.

---

## 🧪 Tests

Added to `src/compat/predicate/isMatch.spec.ts`.

1. Primitive target with object source (Bug 1)

```ts
it('returns false when target is primitive and source has object keys (Issue #1399)', () => {
  expect(isMatch('bar', { anyKey: undefined })).toBe(false);
  expect(isMatch(123, { anyKey: undefined })).toBe(false);
  expect(isMatch(true, { anyKey: undefined })).toBe(false);
});
```

2. Nested empty object matching (Bug 2)

```ts
it('returns false when nested target is primitive and source is empty object (Issue #1399)', () => {
  expect(isMatch({ value: 'bar' }, { value: {} })).toBe(false);
  expect(isMatch({ value: 123 }, { value: {} })).toBe(false);
  expect(isMatch({ value: true }, { value: {} })).toBe(false);
});
```

Result:

- ✅ 25/25 tests passed
- ✅ No type errors

---

## 📊 Summary of Changes

Modified files:

- `src/compat/predicate/isMatchWith.ts` (logic updates)
- `src/compat/predicate/isMatch.spec.ts` (new test cases)

Code added (conceptually):

```ts
// 1) Empty-object handling by level (top-level vs nested)
if (keys.length === 0) {
  if (stack && stack.size > 0 && !isObject(target)) {
    return false;
  }
  return true;
}

// 2) Primitive target cannot match object pattern
if (!isObject(target)) {
  return false;
}
```

Only a few lines, but fully aligns with Lodash behavior.

---

## 🧭 Behavior vs Lodash

Before (incorrect):

```ts
isMatch('bar', { anyKey: undefined }); // ES-Toolkit: true ❌, Lodash: false
isMatch({ value: 'bar' }, { value: {} }); // ES-Toolkit: true ❌, Lodash: false
```

After (fixed):

```ts
isMatch('bar', { anyKey: undefined }); // ES-Toolkit: false ✅, Lodash: false
isMatch({ value: 'bar' }, { value: {} }); // ES-Toolkit: false ✅, Lodash: false
isMatch('bar', {}); // ES-Toolkit: true  ✅, Lodash: true
```

---

## ✅ Checklist

- [x] Fix both issues in #1399
- [x] Match Lodash behavior at both top-level and nested levels
- [x] Add unit tests covering the two bug scenarios
- [x] Keep changes minimal (constant-time checks only)
- [x] Preserve performance; no extra traversal or allocations

raon0211 commented on src/compat/predicate/isMatchWith.ts:
```diff
@@ -211,9 +211,16 @@ function isObjectMatch(
+  if (!isObject(target)) {
+    return false;
+  }
+
```
```suggestion
  if (!isObject(target)) {
    return false;
  }

  if (keys.length === 0) {
    return true;
  }

```

sen2y commented:
> Thanks for your work! Let me refactor this a little bit after merging.

Thanks for reviewing and refining the code! 🙌  
It’s much clearer thanks to your help.

While working on this fix, I also discovered some behavior differences between Lodash and compat. 
I opened an issue to discuss potential improvements and would love your thoughts when you have a moment: https://github.com/toss/es-toolkit/issues/1504

Thanks again — please let me know if there's anything else I can help with before merging! 😊

raon0211 committed:
fix(isMatch, isMatchWith): match lodash for object patterns against primitive and non-plain-object targets

At nested levels, lodash rejects plain-object source patterns (including
empty ones) when the target is not a plain object-like value, using a tag
comparison. At the top level, lodash wraps primitive targets with Object()
so that patterns like { length: 3 } can match 'bar'.

Align isMatchWith with both behaviors:

- isMatch({ value: 'bar' }, { value: {} }) now returns false (was true)
- isMatch({ a: [] }, { a: {} }) now returns false (was true)
- isMatch('bar', { length: 3 }) now returns true (was false)
- isMatch('bar', { anyKey: undefined }) still returns false

Co-authored-by: sen2y <pse6704@sookmyung.ac.kr>

raon0211 reviewed (dismissed):
Thanks!

raon0211 committed:
test: update bundle size snapshots

raon0211 reviewed (approved):
Thanks!
